src/hash.c

```diff
@@ -46,62 +49,157 @@ void EVP_MD_CTX_free(EVP_MD_CTX *ctx)
+	context->error = 0;
+	context->digest_length = 0;
+	context->algo = algo;
+
+	context->context = EVP_MD_CTX_new();
+	if (!context->context)
+		return hash_error(context, ENOMEM, "Unable to allocate hash context.");
+
+	if (!EVP_DigestInit_ex(context->context, context->algo, NULL)) {
+		EVP_MD_CTX_free(context->context);
+		return hash_error(context, ENOTRECOVERABLE, "Unable to initialize hash context.");
+	}
+
+	context->state = HS_UPDATE;
+	return 0;
 }
 
-void module_sha_update(struct image *image, uint8_t *data, size_t bytes)
+int hash_sha256_init(struct hash_context *context)
 {
-	EVP_DigestUpdate(image->mdctx, data, bytes);
+	return hash_init(context, EVP_sha256());
 }
 
-void module_sha_complete(struct image *image, uint8_t *hash)
+int hash_sha384_init(struct hash_context *context)
 {
-	unsigned char md_value[EVP_MAX_MD_SIZE];
-	unsigned int md_len;
-#if DEBUG_HASH
-	int i;
-#endif
-	EVP_DigestFinal_ex(image->mdctx, md_value, &md_len);
-	EVP_MD_CTX_free(image->mdctx);
+	return hash_init(context, EVP_sha384());
+}
+
+int hash_update(struct hash_context *context, const void *data, size_t size)
+{
+	assert(context);
+	assert(data);
+
+	if (context->error)
+		return context->error;
+
+	assert(context->state == HS_UPDATE);
+	
+	if (!EVP_DigestUpdate(context->context, data, size))
+		return hash_error(context, EINVAL, "Unable to update hash context.");
+
+	return 0;
+}
+
+int hash_finalize(struct hash_context *context)
+{
+	assert(context);
+
+	if (context->error)
+		return context->error;
+
+	assert(context->state == HS_UPDATE);
+
+	if (!EVP_DigestFinal_ex(context->context, context->digest, &context->digest_length))
+		return hash_error(context, EINVAL, "Unable to finalize hash context.");
+
+	context->state = HS_DONE;
 
-	memcpy(hash, md_value, md_len);
 #if DEBUG_HASH
-	fprintf(stdout, "Module digest is: ");
-	for (i = 0; i < md_len; i++)
-		fprintf(stdout, "%02x", md_value[i]);
-	fprintf(stdout, "\n");
+	fprintf(stdout, "Hash result is: ");
+	hash_print(context);
 #endif
+
+	hash_free(context);
+	return 0;
 }
 
-void ri_sha256(struct image *image, unsigned int offset, unsigned int size,
-	       uint8_t *hash)
+int hash_get_digest(struct hash_context *context, void *output, size_t output_len)
 {
-	assert((uint64_t)size + offset <= image->adsp->image_size);
-	module_sha256_create(image);
-	module_sha_update(image, image->fw_image + offset, size);
-	module_sha_complete(image, hash);
+	assert(context);
+	assert(output);
+
+	if (context->error)
+		return context->error;
+
+	assert(context->state == HS_DONE);
+
+	if (context->digest_length > output_len)
+		return -ENOBUFS;
+
+	memcpy(output, context->digest, context->digest_length);
+	return context->digest_length;
 }
 
-void module_sha384_create(struct image *image)
+void hash_free(struct hash_context *context)
 {
-	image->md = EVP_sha384();
-	image->mdctx = EVP_MD_CTX_new();
+	assert(context);
+
+	if (context->context)
+		EVP_MD_CTX_free(context->context);
+
+	context->context = NULL;
+}
```

Comment:
I would delete this `hash_free` function since `EVP_MD_CTX_free` does a null check inside its implementation,
Just call inline `EVP_MD_CTX_free(context->context)` everywhere, and then context->context = NULL;
They use it all the time like this in the [OpenSSL 1.1.1](https://github.com/openssl/openssl/tree/OpenSSL_1_1_1-stable) branch and newer 3.0+.
Also it does not make sense to keep 2 line function named "hash_free" that assigns a NULL to pointer freed, since
lexically this is not related to freeing the memory operation. Thus, I would just remove this function.

Reply:
@aborisovich do you agree now?